### PR TITLE
Document Outline: Implement the document outline as a plugin

### PR DESCRIPTION
**Pull request by youknowriad**
refs #4287 

This PR rewrites the document outline as a sidebar plugin. The original intent of the issue was to use a "PopoverPlugin" instead of "SidbarPlugin".

If we don't want to use the SidebarPlugin for now as the PopoverPlugin, This PR is still useful to help build the PopoverPlugin extensibility slot.

One issue I see with the PopoverPlugin though, is that it has no existence without an opening menu item as opposed to the sidebar. Also, where do you show the popover if the menu item is not pinned?

screen shot 2018-04-19 at 12 09 40

### CODE REVIEW

**Comment by gziolo** [comment]
Not that much work to convert it to a plugin. Do we want to do the same for `Document Settings` and `Block Setting`? I didn't take that step, but if we are fine with using plugins mechanism for core features I can open PR to align with what this PR does. 

`Popover` vs `Sidebar` is a design decision. Indeed, opening popover through More Menu can cause some UI challenges.

**Comment by jasmussen** [comment]
Thank you for working on this. Love intent. I would want Matías to be allowed to chime in with a final word, as I know that this is close to his heart. 

In this case, I see two major improvements right off the bat. This area feels much more intentional:

screen shot 2018-04-19 at 13 43 49

It features global actions — to add to the document, or to go back and forth. Previously it felt a bit imbalanced, both optically and in that it was a mix of actions:

screen shot 2018-04-19 at 13 46 00

@gziolo brings up a good point, though, about this being a sidebar, and a popover being a bit weird to open from the more menu. I would note that this is an extra reason why it's healthy to open this PR, so that we can refine those interactions, and test our ideas about how extensibility should be. Right now this is in part modeled after Chrome extensions:

chrome extensions

In our case, it feels like we have a few missing pieces:

- We don't yet have pinning and unpinning of extension icons. This would be a pinned-by-default extension
- The ability to, sidebar open or not, quickly click the little (i) and see an overview, feels good. So popover would be good for this.
- We don't yet have the Star button to pin or unpin an icon to the toolbar, [as mocked up here](https://github.com/WordPress/gutenberg/issues/4287#issuecomment-382367609)

The way popovers could work in the ellipsis menu is that they simply pop up as little popups (with no triangle connector) until you pin them or dismiss them. Similar to the Chrome example above. This is sort of a standard pattern for other apps as well. Here's what it looks like when you open the keyboard shortcuts in Google Docs:

keyboard shortcuts

In fact it would be nice if we could write a keyboard shortcuts tool just like this, that would work in the same way. Pinnable or not. 

---

So to summarize:

- Popover would be good
- Good to wait for pinning feature
- Popup window when invoked from More menu, Pop-out window when invoked from its pinned position in the toolbar.

**Comment by gziolo** [comment]
> So to summarize:
> 
> * Popover would be good
> * Good to wait for pinning feature
> * Popup window when invoked from More menu, Pop-out window when invoked from its pinned position in the toolbar.

+ Pinned by default

It all sounds like a great plan 💯 I'm starting on pinning feature tomorrow, thanks for sharing all details how it is supposed to work 👍

**Comment by SuperGeniusZeb** [comment]
Is this still blocked, and if so, by what?

**Comment by gziolo** [comment]
@jasmussen, I think this is no longer blocked since we have `Modal` component. Is that right?

**Comment by jasmussen** [comment]
> I think this is no longer blocked since we have Modal component. Is that right?

That depends — is the modal able to show as a "popover"?

While I'm personally okay with the document outline becoming a true modal with a close button (though I'd want us to be able to show it more compact than it is now), I think @mtias was dreaming of a version of this that could behave identically to how it works now, only as a plugin.

**Comment by jasmussen** [comment]
An alternative to converting it to a plugin, is to just move the icon over there on the right, between the cog and the ellipsis, even if it isn't a plugin.

**Comment by ZebulanStanphill** [comment]
@jasmussen I think it should be a plugin at some point in the long run, just because it makes more sense to me from a technical perspective, and would provide an example in core of a simple plugin. It would allow you to unpin it from the toolbar if desired.

That said, I do think it makes more sense being on the right than on the left anyway.

**Comment by youknowriad** [comment]
Closing, we might want to revisit at some point.
